File: backend/skullking/src/main/java/com/example/skullking/entities/Room.java
package com.example.skullking.entities;

import com.example.skullking.entities.game.GameState;
import com.example.skullking.entities.game.GameStateMachine;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class Room {

    private UUID uuid;
    private String name;

    private Player host;
    private List<Player> guestsList;


    private GameState gameState;


    public Room() {
        this.uuid = UUID.randomUUID();
        this.host = new Player();
        this.guestsList = new ArrayList<>();
        this.name = "";
        this.gameState = new GameState();
    }

    public UUID getUuid() {
        return this.uuid;
    }

    public void addGuest(Player player) {

        if (!this.guestsList.contains(player)) {

            this.guestsList.add(player);
        }
    }

    public void deleteGuest(UUID playerUuid) {

        this.guestsList.removeIf(p -> p.getUuid().equals(playerUuid));
    }

    public int countPlayers() {

        // the host + the guests
        return 1 + this.guestsList.size();
    }

    public void setHost(Player host) {
        this.host = host;
    }

    public Player getHost() {
        return this.host;
    }

    public List<Player> getPlayers() {

        ArrayList<Player> copiedList = new ArrayList<>(this.guestsList);
        copiedList.add(this.host);

        return copiedList;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }


    public GameState getGameState() {
        return this.gameState;
    }
}

File: backend_py/app/entities/room.py
import uuid
from typing import List

from app.entities.player import Player
from app.entities.game import Game


class Room:
    def __init__(self, name: str, host: Player):
        self.uuid: uuid.UUID = uuid.uuid4()
        self.name: str = name

        self.host: Player = host
        self.guests_list: List[Player] = []

        self.game: Game = Game()

    def add_guest(self, player: Player) -> None:
        if player not in self.guests_list:
            self.guests_list.append(player)

    def delete_guest(self, player_uuid: uuid.UUID) -> None:
        self.guests_list = [
            p for p in self.guests_list if p.uuid != player_uuid
        ]

    def count_players(self) -> int:
        return 1 + len(self.guests_list)
    
    def players(self)-> List[Player]:
        return [self.host] + self.guests_list


    def get_players(self) -> List[Player]:
        # on retourne une COPIE pour éviter les effets de bord
        players = self.guests_list.copy()
        players.append(self.host)
        return players

    def get_player_by_token(self, token: str) -> Player | None:
        if self.host.token == token:
            return self.host
        for guest in self.guests_list:
            if guest.token == token:
                return guest
        return None
    
    def get_player_by_uuid(self, player_uuid: uuid.UUID) -> Player | None:
        if self.host.uuid == player_uuid:
            return self.host
        for guest in self.guests_list:
            if guest.uuid == player_uuid:
                return guest
        return None
